server/src/main/java/org/elasticsearch/reservedstate/service/AbstractFileWatchingService.java

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.reservedstate.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.support.PlainActionFuture;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Randomness;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.env.Environment;
+
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.FileTime;
+import java.time.Instant;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * <p>A file watching service watches for changes in a particular file on disk. There
+ * are three assumptions about the file structure:</p>
+ * <ol>
+ *     <li>The file itself may or may not exist.</li>
+ *     <li>The file's parent directory may or may not exist.</li>
+ *     <li>The directory above the file's parent directory must always exist.</li>
+ * </ol>
+ *
+ * <p>For example, if the watched file is under /usr/elastic/elasticsearch/conf/special/settings.yml,
+ * then /usr/elastic/elasticsearch/conf/ must exist, but special/ and special/settings.yml may
+ * be created, updated, or deleted during runtime.</p>
+ *
+ * <p>What this class does not do is define what should happen after the file changes.
+ * An implementation class should override {@link #processFileChanges(Path)} to define
+ * the correct behavior.</p>
+ */
+public abstract class AbstractFileWatchingService extends AbstractLifecycleComponent implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(FileSettingsService.class);
+    protected static final int REGISTER_RETRY_COUNT = 5;
```

Comment:
could most (or all?) of these members be private?

Reply:
Yes, they can. I used intelliJ's refactoring and failed to clean up access levels afterwards.